src/nbl/video/CVulkanConnection.h

```diff
@@ -42,79 +59,86 @@ class CVulkanConnection final : public IAPIConnection
+        {
+            printf("Failed to create debug messenger for some reason!\n");
+            return nullptr;
+        }
 
-                instanceCreateInfo.enabledExtensionCount = instanceExtensionCount;
-                instanceCreateInfo.ppEnabledExtensionNames = instanceExtensions;
-            }
-            else
+        constexpr uint32_t MAX_PHYSICAL_DEVICE_COUNT = 16u;
+        uint32_t physicalDeviceCount = 0u;
+        VkPhysicalDevice vk_physicalDevices[MAX_PHYSICAL_DEVICE_COUNT];
+        {
+            VkResult retval = vkEnumeratePhysicalDevices(vk_instance, &physicalDeviceCount, nullptr);
+            if ((retval != VK_SUCCESS) && (retval != VK_INCOMPLETE))
             {
-                // Todo(achal): This needs to be handled in a platform agnostic way.
-                const uint32_t instanceExtensionCount = 3u;
-                char* instanceExtensions[instanceExtensionCount] = { "VK_KHR_surface", "VK_KHR_win32_surface", VK_EXT_DEBUG_UTILS_EXTENSION_NAME };
-
-                instanceCreateInfo.enabledExtensionCount = instanceExtensionCount;
-                instanceCreateInfo.ppEnabledExtensionNames = instanceExtensions;
+                printf("Failed to enumerate physical devices!\n");
+                return nullptr;
             }
+
+            vkEnumeratePhysicalDevices(vk_instance, &physicalDeviceCount, vk_physicalDevices);
         }
-        
-        vkCreateInstance(&instanceCreateInfo, nullptr, &m_instance);
-        assert(m_instance);
-        
-        // Todo(achal): I gotta look into this ie if (and how) we want to use volkLoadInstanceOnly
-        // volkLoadInstanceOnly(m_instance);
-        volkLoadInstance(m_instance);
-
-        vkCreateDebugUtilsMessengerEXT(m_instance, &debugUtilsMessengerCreateInfo, nullptr, &m_debugMessenger);
-        assert(m_debugMessenger);
-        
-        uint32_t devCount = 0u;
-        vkEnumeratePhysicalDevices(m_instance, &devCount, nullptr);
-        core::vector<VkPhysicalDevice> vkphds(devCount, VK_NULL_HANDLE);
-        vkEnumeratePhysicalDevices(m_instance, &devCount, vkphds.data());
-
-        m_physDevices = core::make_refctd_dynamic_array<physical_devs_array_t>(devCount);
-        for (uint32_t i = 0u; i < devCount; ++i)
+
+        physical_devs_array_t physicalDevices = core::make_refctd_dynamic_array<physical_devs_array_t>(physicalDeviceCount);
+        for (uint32_t i = 0u; i < physicalDeviceCount; ++i)
         {
-            (*m_physDevices)[i] = core::make_smart_refctd_ptr<CVulkanPhysicalDevice>(vkphds[i], core::smart_refctd_ptr(m_system), std::move(m_GLSLCompiler));
+            (*physicalDevices)[i] = core::make_smart_refctd_ptr<CVulkanPhysicalDevice>(
+                vk_physicalDevices[i], core::smart_refctd_ptr(sys),
+                core::make_smart_refctd_ptr<asset::IGLSLCompiler>(sys.get()));
         }
+
+        return core::make_smart_refctd_ptr<CVulkanConnection>(vk_instance, physicalDevices, vk_debugMessenger);
```

Comment:
the trick with creating a class with a protected constructor is to do this
```cpp
auto retval = new CVulkanConnection(vk_instance,physicalDevices,std::move(callback));
return core::smar_refctd_ptr<CVulkanConnection>(retval,core::dont_grab);
```

Reply:
This is because the `core::make_smart_refctd_ptr` cannot call a `proctected`/`private` constructor, so you call the constructor in the class' `create` function --essentially inlining the impl of `core::make_smart_refctd_ptr`.

Reply:
Resolved